File: tools/firebase_import.js
/**
 * Firebase Firestore Import Script
 * 
 * Imports inventory data to Firestore using Firebase Admin SDK
 * 
 * Prerequisites:
 *   1. npm install firebase-admin
 *   2. Set GOOGLE_APPLICATION_CREDENTIALS environment variable
 *      OR ensure gcloud is authenticated
 * 
 * Usage:
 *   node tools/firebase_import.js
 */

const admin = require('firebase-admin');
const fs = require('fs');
const path = require('path');

// Initialize Firebase Admin
// Uses application default credentials (gcloud auth)
admin.initializeApp({
  projectId: 'sapphire-479610'
});

const db = admin.firestore();
const COLLECTION = 'inventory';

async function importInventory() {
  // Read the import data
  const dataPath = path.join(__dirname, '..', 'data', 'firestore_inventory_import.json');
  const rawData = fs.readFileSync(dataPath, 'utf8');
  const inventory = JSON.parse(rawData);
  
  const docIds = Object.keys(inventory);
  console.log(`Found ${docIds.length} inventory items to import\n`);
  
  let imported = 0;
  let updated = 0;
  let errors = [];
  
  for (const docId of docIds) {
    try {
      const data = inventory[docId];
      const docRef = db.collection(COLLECTION).doc(docId);
      
      // Check if document exists
      const existing = await docRef.get();
      
      if (existing.exists) {
        await docRef.update(data);
        console.log(`✓ Updated: ${data.model_name}`);
        updated++;
      } else {
        await docRef.set(data);
        console.log(`✓ Created: ${data.model_name}`);
        imported++;
      }
    } catch (error) {
      console.error(`✗ Error importing ${docId}:`, error.message);
      errors.push({ id: docId, error: error.message });
    }
  }
  
  console.log('\n' + '='.repeat(60));
  console.log('IMPORT SUMMARY');
  console.log('='.repeat(60));
  console.log(`Created: ${imported}`);
  console.log(`Updated: ${updated}`);
  console.log(`Errors:  ${errors.length}`);
  console.log('='.repeat(60));
  
  if (errors.length > 0) {
    console.log('\nErrors encountered:');
    errors.forEach(e => console.log(`  - ${e.id}: ${e.error}`));
  }
  
  process.exit(errors.length > 0 ? 1 : 0);
}

// Run the import
importInventory().catch(error => {
  console.error('Import failed:', error);
  process.exit(1);
});

File: tools/firebase_import.py
"""
Firebase Firestore Import Script

Imports inventory data to Firestore using Google Cloud Firestore client.

Prerequisites:
    pip install google-cloud-firestore

Usage:
    python tools/firebase_import.py
"""

import json
import os
import sys
from pathlib import Path

# Add parent directory to path
sys.path.insert(0, str(Path(__file__).parent.parent))

try:
    from google.cloud import firestore
except ImportError:
    print("Error: google-cloud-firestore not installed.")
    print("Run: pip install google-cloud-firestore")
    sys.exit(1)


def import_inventory():
    """Import inventory to Firestore."""
    # Initialize Firestore client
    project_id = os.environ.get("FIRESTORE_PROJECT_ID", "sapphire-479610")
    db = firestore.Client(project=project_id)
    collection = "inventory"

    # Read import data
    data_path = Path(__file__).parent.parent / "data" / "firestore_inventory_import.json"
    with open(data_path) as f:
        inventory = json.load(f)

    doc_ids = list(inventory.keys())
    print(f"Found {len(doc_ids)} inventory items to import\n")

    imported = 0
    updated = 0
    errors = []

    for doc_id in doc_ids:
        try:
            data = inventory[doc_id]
            doc_ref = db.collection(collection).document(doc_id)

            # Check if document exists
            existing = doc_ref.get()

            if existing.exists:
                doc_ref.update(data)
                print(f"✓ Updated: {data['model_name']}")
                updated += 1
            else:
                doc_ref.set(data)
                print(f"✓ Created: {data['model_name']}")
                imported += 1

        except Exception as e:
            print(f"✗ Error importing {doc_id}: {e}")
            errors.append({"id": doc_id, "error": str(e)})

    print("\n" + "=" * 60)
    print("IMPORT SUMMARY")
    print("=" * 60)
    print(f"Created: {imported}")
    print(f"Updated: {updated}")
    print(f"Errors:  {len(errors)}")
    print("=" * 60)

    if errors:
        print("\nErrors encountered:")
        for e in errors[:10]:
            print(f"  - {e['id']}: {e['error']}")
        if len(errors) > 10:
            print(f"  ... and {len(errors) - 10} more")

    return len(errors) == 0


if __name__ == "__main__":
    success = import_inventory()
    sys.exit(0 if success else 1)
